### PR TITLE
feat: add search over all filters (DataBiosphere/findable-ui#13)

### PR DESCRIPTION
**Pull request by hunterckx**
Not implemented:
- Match highlighting
- Prevention of scrolling outside the menu
- Tall menu (currently the menu is the same hight as the normal filter menus)
- Analytics event

### CODE REVIEW

File: packages/data-explorer-ui/src/components/Filter/components/FilterNoResultsFound/filterNoResultsFound.tsx [hunk 1]
```diff
@@ -20,15 +20,19 @@ export const FilterNoResultsFound = ({
...
+        mb={onClearSearchTerm ? 2 : 0}
         mt={1}
         variant="text-body-400"
       >
         Try adjusting your search or filter to find what you’re looking for.
       </Typography>
-      <TextButtonPrimary onClick={onClearSearchTerm}>
-        Clear All
-      </TextButtonPrimary>
+      {onClearSearchTerm ? (
```
Comment (frano-m) [inline packages/data-explorer-ui/src/components/Filter/components/FilterNoResultsFound/filterNoResultsFound.tsx:29]: @hunterckx another possible way to write lines `29-35` would be:

```
{onClearSearchTerm && <TextButtonPrimary onClick={onClearSearchTerm}>Clear All</TextButtonPrimary>}
```

File: packages/data-explorer-ui/src/components/Filter/components/FilterNoResultsFound/filterNoResultsFound.tsx [hunk 2]
```diff
@@ -20,15 +20,19 @@ export const FilterNoResultsFound = ({
...
+        <TextButtonPrimary onClick={onClearSearchTerm}>
+          Clear All
+        </TextButtonPrimary>
+      ) : (
+        ""
```
Comment (frano-m) [inline packages/data-explorer-ui/src/components/Filter/components/FilterNoResultsFound/filterNoResultsFound.tsx:34]: If `onClearSearchTerm` is not defined, we can return `null`, so an alternative to https://github.com/clevercanary/data-explorer/pull/280/files#r1267366143 would be:

```
{onClearSearchTerm ? (<TextButtonPrimary onClick={onClearSearchTerm}>Clear All</TextButtonPrimary>) : null}
```

File: packages/data-explorer-ui/src/components/Filter/components/Filters/filters.tsx [hunk 3]
```diff
@@ -84,17 +85,20 @@ export const Filters = ({
...
-          key={categoryView.key}
-          tags={renderFilterTags(categoryView, onFilter)}
-          Target={(props): JSX.Element =>
-            renderFilterTarget(categoryView, props)
-          }
-        />
-      ))}
-    </FilterList>
+    <>
+      <SearchAllFilters categoryViews={categoryViews} onFilter={onFilter} />
```
Comment (frano-m) [inline packages/data-explorer-ui/src/components/Filter/components/Filters/filters.tsx:89]: For flexibility of components, I think we should move the render of `SearchAllFilters` to https://github.com/clevercanary/data-explorer/blob/5376ef153faab249c6da4dd3e055b926729ad40c/packages/data-explorer-ui/src/views/ExploreView/exploreView.tsx#L109 (above the render of the `Filters` component)

File: packages/data-explorer-ui/src/components/Filter/components/SearchAllFiltersMenu/searchAllFiltersMenu.styles.ts
Comment (frano-m) [inline packages/data-explorer-ui/src/components/Filter/components/SearchAllFiltersMenu/searchAllFiltersMenu.styles.ts:1]: The styles for `FilterView` look to be the same as the `FilterView` from https://github.com/clevercanary/data-explorer/blob/5376ef153faab249c6da4dd3e055b926729ad40c/packages/data-explorer-ui/src/components/Filter/components/FilterMenu/filterMenu.styles.ts. Let's import this `FilterView` instead?

File: packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.tsx [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  if (!inputValue) return options;
+  inputValue = inputValue.toLowerCase();
+  const filteredOptions = options.filter(
+    (option) =>
+      typeof option !== "string" &&
+      (option.value.key?.toLowerCase().includes(inputValue) ||
+        option.value.label?.toLowerCase().includes(inputValue))
+  );
+  // noOptionsText isn't supported for freeSolo Autocomplete, so simulate with a special option value
+  if (filteredOptions.length === 0) return [SPECIAL_OPTION.NO_RESULTS];
```
Comment (frano-m) [inline packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.tsx:171]: NICE 🚀

File: packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.tsx [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+      PaperComponent={PaperComponent}
+      PopperComponent={PopperComponent}
+      renderGroup={renderGroup}
+      renderInput={renderInput}
+      renderOption={renderOption}
+    />
+  );
+};
+
+function applyMenuFilter(
```
Comment (frano-m) [inline packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.tsx:158]: Fantastic function!

File: packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.tsx [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+import {
+  GroupDivider,
+  GroupHeading,
+  SearchAllFilters as Autocomplete,
+  SearchAllFiltersMenuPaper,
+} from "./searchAllFilters.styles";
+
+// Special option values, for insterting menu content other than actual options
+enum SPECIAL_OPTION {
+  NO_RESULTS = "SPECIAL_OPTION_NO_RESULTS",
```
Comment (frano-m) [inline packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.tsx:32]: LOVE!

File: packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.tsx [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+  return (
+    <Autocomplete
+      filterOptions={applyMenuFilter}
+      freeSolo
+      getOptionLabel={(option): string =>
+        typeof option === "string" ? option : option.value.label
+      }
+      groupBy={(option): string =>
+        typeof option === "string" ? "" : option.categoryLabel
```
Comment (frano-m) [inline packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.tsx:145]: 🚀

File: packages/data-explorer-ui/src/components/Filter/components/SearchAllFiltersMenu/searchAllFiltersMenu.styles.ts
Comment (hunterckx) [reply packages/data-explorer-ui/src/components/Filter/components/SearchAllFiltersMenu/searchAllFiltersMenu.styles.ts:1]: Would this conflict with the filter menu virtualization? My inclination would be to keep them separate until this is virtualized as well

File: packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.tsx [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+);
+
+const renderGroup = ({
+  children,
+  group,
+  key,
+}: AutocompleteRenderGroupParams): JSX.Element =>
+  group ? (
+    <Fragment key={key}>
+      <GroupDivider />
```
Comment (frano-m) [inline packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.tsx:74]: Try: 
```
{Number(key) !== 0 && <GroupDivider />}
```
Not sure why `AutocompleteRenderGroupParams` returns key as a string? Convert to number and test that the first key is not equal to 0...

File: packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.styles.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+`;
+
+export const GroupHeading = styled.div`
+  padding: 8px 16px;
+`;
+
+export const GroupDivider = styled(Divider)`
+  margin: 8px 0;
+  &&:first-of-type {
+    display: none;
```
[the remark this answers is not in the document]
Comment (frano-m) [reply packages/data-explorer-ui/src/components/Filter/components/SearchAllFilters/searchAllFilters.styles.ts:23]: See https://github.com/clevercanary/data-explorer/pull/280/files#r1267422370. Let me know what you think?

File: packages/data-explorer-ui/src/components/Filter/components/SearchAllFiltersMenu/searchAllFiltersMenu.styles.ts
Comment (frano-m) [reply packages/data-explorer-ui/src/components/Filter/components/SearchAllFiltersMenu/searchAllFiltersMenu.styles.ts:1]: Ok, let's leave it for now! Good suggestion!

File: packages/data-explorer-ui/src/components/Filter/components/SearchAllFiltersSearch/searchAllFiltersSearch.tsx [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    <InputField
+      {...props}
+      variant="outlined"
+      fullWidth
+      InputProps={{
+        ...(InputProps || {}),
+        startAdornment: <SearchIcon fontSize="small" />,
+      }}
+      inputProps={{ ...(inputProps || {}), spellCheck: false }}
+      placeholder="Search all filters"
```
Comment (frano-m) [inline packages/data-explorer-ui/src/components/Filter/components/SearchAllFiltersSearch/searchAllFiltersSearch.tsx:21]:
```suggestion
      placeholder="Search all filters..."
```

**Review by frano-m — APPROVED** [review]
Outstanding.

**Pull request merged**